### PR TITLE
Add fuse mount support

### PR DESCRIPTION
**Pull request by bolkedebruin**
This add support for /dev/fuse which allows users to mount fuse file systems. "mount" syscall support is required (seccomp)

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

**Comment by bolkedebruin** [comment]
cc @phoracek can yoiu have a look please?

**Comment by SchSeba** [comment]
Hi @bolkedebruin can you please add more details on the use-case for this feature?

**Comment by bolkedebruin** [comment]
@SchSeba You may want to give users the possibility to mount fuse filesystems (s3, gcs, sshfs) in a pod. Fuse filesystems are userspace and thus root isnt required, however they require the special device /dev/fuse . This device is only available when you use "priviliged" pods (and provided by docker), which defeats the purpose and is not allowed in secure deployments. Hence having a DaemonSet that provides '/dev/fuse' to pods that request it.

**Comment by phoracek** [comment]
@bolkedebruin could you please add some documentation https://github.com/kubevirt/kubernetes-device-plugins/tree/master/docs?

**Comment by bolkedebruin** [comment]
Absolutely. Will do

**Comment by phoracek** [comment]
@bolkedebruin great :) while at it, don't forget to update the main README https://github.com/kubevirt/kubernetes-device-plugins#current-device-plugins

**Comment by bolkedebruin** [comment]
@phoracek done.

File: docs/README.fuse.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    ...
+    resources:
+      requests:
+              devices.kubevirt.io/fuse: "1"
+      limits:
+              devices.kubevirt.io/fuse: "1"
+```
+
+The device does not need a priviliged container to use it, but it does require the `mount` system call and CAP_SETUID. The `mount` system call can be made
+available by a seccomp profile.
```
Comment (phoracek) [inline docs/README.fuse.md]: how do i create a seccomp profile?

File: docs/README.fuse.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+## Issues
+
+### FUSE Quantity
+
+Currently, Kubernetes device plugins can only expose quantifiable resources.
+FUSE itself is "infinite" resource, so it cannot be properly expressed.
+
+This fact is worked around by the plugin: it adds one more FUSE resource each
+time VM is started. Because we have no way of detecting stopped VMs, the FUSE
```
Comment (phoracek) [inline docs/README.fuse.md:82]: VM is started? Is this device plugin meant to be used from KubeVirt? How would it look like?

File: docs/README.fuse.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+available by a seccomp profile.
+
+## Issues
+
+### FUSE Quantity
+
+Currently, Kubernetes device plugins can only expose quantifiable resources.
+FUSE itself is "infinite" resource, so it cannot be properly expressed.
+
+This fact is worked around by the plugin: it adds one more FUSE resource each
```
Comment (phoracek) [inline docs/README.fuse.md:81]: This approach might be raceful. Second VM could request the resource before the number is updated. Consider changing it to hardcoded high-enough amount. There is only limited number of pods you can run on a node anyways.

File: pkg/fuse/fuse.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+// object responsible for discovering initial pool of devices and their allocation.
+type FuseLister struct{}
+
+func (l FuseLister) GetResourceNamespace() string {
+	return Namespace
+}
+
+// Discovery discovers all FUSE devices within the system.
```
Comment (phoracek) [inline pkg/fuse/fuse.go]: This comment is missleading, it exposes single FUSE device from a predefined path. Please correct me if I'm wrong

**Review by phoracek — CHANGES_REQUESTED** [review]

File: docs/README.fuse.md (hunk 3 above)
Comment (bolkedebruin) [reply docs/README.fuse.md:81]: This is equal to the KVM device plugin, after which I modeled this). The number is actually not important really, but it needs to be different for kubernetes to accept it.

File: docs/README.fuse.md (hunk 2 above)
Comment (bolkedebruin) [reply docs/README.fuse.md:82]: e.g. s3fs mybucket /path/to/mountpoint -o passwd_file=${HOME}/.passwd-s3fs -o url=https://url.to.s3/ -o use_path_request_style

File: docs/README.fuse.md (hunk 3 above)
Comment (bolkedebruin) [reply docs/README.fuse.md:81]: I can change it to an UUID, but not sure if it brings much

File: docs/README.fuse.md (hunk 2 above)
Comment (phoracek) [reply docs/README.fuse.md:82]: Please add an example to documentation.

File: docs/README.fuse.md (hunk 3 above)
Comment (phoracek) [reply docs/README.fuse.md:81]: I don't believe the KVM DP from this repo is used in KubeVirt.

@stu-gott could you please tell us how does the KVM DP handle its pool of "devices"? Does it use static number of available devices or does it dynamically increase it?

File: docs/README.fuse.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+The device does not need a priviliged container to use it, but it does require the `mount` system call and CAP_SETUID. The `mount` and `umount` system calls can be made available by a [security context (seccomp) profile](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/).
+
+## Issues
+
+### FUSE Quantity
+
+Currently, Kubernetes device plugins can only expose quantifiable resources.
+FUSE itself is "infinite" resource, so it cannot be properly expressed.
+
+This fact is worked around by the plugin: it adds one more FUSE resource each
```
Comment (phoracek) [inline docs/README.fuse.md:81]: outdated

File: pkg/fuse/fuse.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	FUSEPath 		 = "/dev/fuse"
+	FUSEName  		= "fuse"
+	Namespace 		= "devices.kubevirt.io"
+	fusePoolSize 	= 128
+)
+
+type message struct{}
+
+type FusePlugin struct {
+	counter int
```
Comment (phoracek) [inline pkg/fuse/fuse.go:24]: the counter is not needed anymore, is it?

**Review by phoracek — CHANGES_REQUESTED** [review]
Looks good, just 2 comments.

Have you tried this plugin with KubeVirt? Does it work?

File: pkg/fuse/fuse.go (hunk 6 above)
Comment (bolkedebruin) [reply pkg/fuse/fuse.go:24]: No I will remove it

File: docs/README.fuse.md (hunk 5 above)
Comment (bolkedebruin) [reply docs/README.fuse.md:81]: Will fix

**Pull request closed without merging**